// axiodb/lib/config/DB.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.InstanceTypes = exports.AxioDBCloud = exports.AxioDB = void 0;
// Import All Required Sub Modules
const memory_operation_1 = require("../Memory/memory.operation");
const Converter_helper_1 = __importDefault(require("../Helper/Converter.helper"));
const Crypto_helper_1 = require("../Helper/Crypto.helper");
const response_helper_1 = __importDefault(require("../Helper/response.helper"));
const Aggregation_Operation_1 = __importDefault(require("../Services/Aggregation/Aggregation.Operation"));
const collection_operation_1 = __importDefault(require("../Services/Collection/collection.operation"));
const database_operation_1 = __importDefault(require("../Services/Database/database.operation"));
const Indexation_operation_1 = require("../Services/Indexation.operation");
Object.defineProperty(exports, "AxioDB", { enumerable: true, get: function () { return Indexation_operation_1.AxioDB; } });
const AxioDBCloud_client_1 = require("../client/AxioDBCloud.client");
Object.defineProperty(exports, "AxioDBCloud", { enumerable: true, get: function () { return AxioDBCloud_client_1.AxioDBCloud; } });
const FileManager_1 = __importDefault(require("../engine/Filesystem/FileManager"));
const FolderManager_1 = __importDefault(require("../engine/Filesystem/FolderManager"));
// Export All Required Sub Modules Instance Types to support TypeScript
const InstanceTypes = {
    Collection: collection_operation_1.default,
    Database: database_operation_1.default,
    Aggregation: Aggregation_Operation_1.default,
    FileManager: FileManager_1.default,
    FolderManager: FolderManager_1.default,
    Converter: Converter_helper_1.default,
    CryptoHelper: Crypto_helper_1.CryptoHelper,
    ResponseHelper: response_helper_1.default,
    InMemoryCache: memory_operation_1.InMemoryCache,
};
exports.InstanceTypes = InstanceTypes;
// Export With All Sub Modules
exports.default = {
    AxioDB: Indexation_operation_1.AxioDB,
    AxioDBCloud: AxioDBCloud_client_1.AxioDBCloud,
    InstanceTypes,
};
