// mongosuper/lib/Config/Provider.js
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "../Js/Connection/ConnectMongo", "../Js/Service/Create", "../Js/Service/Read", "../Js/Service/Update", "../Js/Service/Delete", "../Ts/Schema/CreateSchema", "../Ts/Model/CreateModel"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.DeleteData = exports.UpdateData = exports.ReadData = exports.CreateData = exports.ConnectDB = void 0;
    const ConnectMongo_1 = __importDefault(require("../Js/Connection/ConnectMongo")); // import MongoDB Connection Function
    // CRUD Functions
    const Create_1 = require("../Js/Service/Create"); // Import Create Function
    const Read_1 = require("../Js/Service/Read"); // Import Read Function
    const Update_1 = require("../Js/Service/Update"); // Import Update Function
    const Delete_1 = require("../Js/Service/Delete"); // Import Delete Function
    // Schema & Models creator
    const CreateSchema_1 = require("../Ts/Schema/CreateSchema"); // Import CreateSchema Function
    const CreateModel_1 = require("../Ts/Model/CreateModel"); // Import CreateModel Function
    /* Export All Service Functions */
    exports.ConnectDB = ConnectMongo_1.default;
    // CRUD Operations
    exports.CreateData = Create_1.Create;
    exports.ReadData = Read_1.Find;
    exports.UpdateData = Update_1.Update;
    exports.DeleteData = Delete_1.Delete;
    // Export With Default Export
    exports.default = Object.freeze({
        CreateModel: CreateModel_1.CreateModel,
        CreateSchema: CreateSchema_1.CreateSchema
    });
});
