// axiodb/lib/client/AggregationProxy.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const command_types_1 = require("../tcp/types/command.types");
/**
 * Aggregation Proxy - Aggregation builder for remote aggregations
 * Mirrors the Aggregation class API
 */
class AggregationProxy {
    constructor(client, dbName, collectionName, pipeline) {
        this.client = client;
        this.dbName = dbName;
        this.collectionName = collectionName;
        this.pipeline = pipeline;
    }
    /**
     * Execute the aggregation pipeline
     */
    exec() {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this.client.sendCommand(command_types_1.CommandType.AGGREGATE, {
                dbName: this.dbName,
                collectionName: this.collectionName,
                pipeline: this.pipeline,
            });
        });
    }
}
exports.default = AggregationProxy;
