// axiodb/lib/tcp/handler/handlers/Collection.handler.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const keys_1 = require("../../config/keys");
const Collection_controller_1 = __importDefault(require("../../../server/controller/Collections/Collection.controller"));
/**
 * Collection Handler - Handles collection-related TCP commands
 * Uses AxioDB instance directly for idempotent operations
 */
class CollectionHandler {
    constructor(axioDB) {
        this.axioDB = axioDB;
        this.controller = new Collection_controller_1.default(axioDB);
    }
    /**
     * Handle CREATE_COLLECTION command
     * Note: This is idempotent - returns collection whether it exists or not
     */
    handleCreateCollection(requestId, params) {
        return __awaiter(this, void 0, void 0, function* () {
            const { dbName, collectionName, crypto, key } = params;
            try {
                // Use AxioDB instance directly for idempotent behavior
                const databaseInstance = yield this.axioDB.createDB(dbName);
                yield databaseInstance.createCollection(collectionName, crypto, key);
                return {
                    id: requestId,
                    statusCode: keys_1.StatusCode.OK,
                    message: 'Collection created successfully',
                    data: { dbName, collectionName },
                };
            }
            catch (error) {
                return {
                    id: requestId,
                    statusCode: keys_1.StatusCode.INTERNAL_SERVER_ERROR,
                    message: error instanceof Error ? error.message : String(error),
                    error: error instanceof Error ? error.message : String(error),
                };
            }
        });
    }
    /**
     * Handle DELETE_COLLECTION command
     */
    handleDeleteCollection(requestId, params) {
        return __awaiter(this, void 0, void 0, function* () {
            const { dbName, collectionName } = params;
            // Create mock request object
            const mockRequest = {
                query: { dbName, collectionName },
            };
            const result = yield this.controller.deleteCollection(mockRequest);
            return {
                id: requestId,
                statusCode: result.statusCode,
                message: result.message,
                data: result.data,
            };
        });
    }
    /**
     * Handle COLLECTION_EXISTS command
     */
    handleCollectionExists(requestId, params) {
        return __awaiter(this, void 0, void 0, function* () {
            const { dbName, collectionName } = params;
            try {
                const databaseInstance = yield this.axioDB.createDB(dbName);
                const exists = yield databaseInstance.isCollectionExists(collectionName);
                return {
                    id: requestId,
                    statusCode: keys_1.StatusCode.OK,
                    message: exists ? 'Collection exists' : 'Collection does not exist',
                    data: { exists },
                };
            }
            catch (error) {
                return {
                    id: requestId,
                    statusCode: keys_1.StatusCode.INTERNAL_SERVER_ERROR,
                    message: error instanceof Error ? error.message : String(error),
                    error: error instanceof Error ? error.message : String(error),
                };
            }
        });
    }
    /**
     * Handle GET_COLLECTION_INFO command
     */
    handleGetCollectionInfo(requestId, params) {
        return __awaiter(this, void 0, void 0, function* () {
            const { dbName } = params;
            // Create mock request object
            const mockRequest = {
                query: { databaseName: dbName },
            };
            const result = yield this.controller.getCollections(mockRequest);
            return {
                id: requestId,
                statusCode: result.statusCode,
                message: result.message,
                data: result.data,
            };
        });
    }
}
exports.default = CollectionHandler;
