// axiodb/lib/tcp/handler/handlers/DB.handler.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const keys_1 = require("../../config/keys");
const Databse_controller_1 = __importDefault(require("../../../server/controller/Database/Databse.controller"));
/**
 * Database Handler - Handles database-related TCP commands
 * Uses AxioDB instance directly for idempotent operations
 */
class DBHandler {
    constructor(axioDB) {
        this.axioDB = axioDB;
        this.controller = new Databse_controller_1.default(axioDB);
    }
    /**
     * Handle CREATE_DB command
     * Note: This is idempotent - returns database whether it exists or not
     */
    handleCreateDB(requestId, params) {
        return __awaiter(this, void 0, void 0, function* () {
            const { dbName } = params;
            try {
                // Use AxioDB instance directly for idempotent behavior
                yield this.axioDB.createDB(dbName);
                return {
                    id: requestId,
                    statusCode: keys_1.StatusCode.OK,
                    message: keys_1.SuccessMessage.DB_CREATED,
                    data: { Database_Name: dbName },
                };
            }
            catch (error) {
                return {
                    id: requestId,
                    statusCode: keys_1.StatusCode.INTERNAL_SERVER_ERROR,
                    message: error instanceof Error ? error.message : String(error),
                    error: error instanceof Error ? error.message : String(error),
                };
            }
        });
    }
    /**
     * Handle DELETE_DB command
     */
    handleDeleteDB(requestId, params) {
        return __awaiter(this, void 0, void 0, function* () {
            const { dbName } = params;
            // Create mock request object
            const mockRequest = {
                query: { dbName },
            };
            const result = yield this.controller.deleteDatabase(mockRequest);
            return {
                id: requestId,
                statusCode: result.statusCode,
                message: result.message,
                data: result.data,
            };
        });
    }
    /**
     * Handle DB_EXISTS command
     */
    handleDBExists(requestId, params) {
        return __awaiter(this, void 0, void 0, function* () {
            const { dbName } = params;
            try {
                const exists = yield this.axioDB.isDatabaseExists(dbName);
                return {
                    id: requestId,
                    statusCode: keys_1.StatusCode.OK,
                    message: exists ? 'Database exists' : 'Database does not exist',
                    data: { exists },
                };
            }
            catch (error) {
                return {
                    id: requestId,
                    statusCode: keys_1.StatusCode.INTERNAL_SERVER_ERROR,
                    message: error instanceof Error ? error.message : String(error),
                    error: error instanceof Error ? error.message : String(error),
                };
            }
        });
    }
    /**
     * Handle GET_INSTANCE_INFO command
     */
    handleGetInstanceInfo(requestId) {
        return __awaiter(this, void 0, void 0, function* () {
            const result = yield this.controller.getDatabases();
            return {
                id: requestId,
                statusCode: result.statusCode,
                message: result.message,
                data: result.data,
            };
        });
    }
}
exports.default = DBHandler;
