// axiodb/lib/server/router/Router.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = mainRouter;
const keys_1 = require("../config/keys");
const responseBuilder_helper_1 = __importDefault(require("../helper/responseBuilder.helper"));
const outers_1 = require("outers");
const promises_1 = require("node:fs/promises");
// All Sub Routers
const DB_routes_1 = __importDefault(require("./Routers/DB.routes"));
const Collection_routes_1 = __importDefault(require("./Routers/Collection.routes"));
const Operation_routes_1 = __importDefault(require("./Routers/Operation.routes"));
const Stats_controller_1 = __importDefault(require("../controller/Stats.controller"));
/**
 * Main router plugin for the AxioDB server
 * @param fastify - Fastify instance
 * @param _options - Plugin options
 * @param done - Callback to signal completion
 */
function mainRouter(fastify, options, done) {
    return __awaiter(this, void 0, void 0, function* () {
        // Now you can access the AxioDB instance
        const { AxioDBInstance } = options;
        fastify.get("/info", () => __awaiter(this, void 0, void 0, function* () {
            const PackageFile = JSON.parse(yield (0, promises_1.readFile)("./package.json", "utf-8"));
            const Reply = (0, responseBuilder_helper_1.default)(outers_1.StatusCodes.OK, "AxioDB Information", {
                Package_Name: PackageFile.name,
                AxioDB_Version: PackageFile.version,
                Author_Name: PackageFile.author,
                License: PackageFile.license,
                AuthorDetails: keys_1.AuthorInfo,
            });
            return Reply;
        }));
        // Health check route
        fastify.get("/health", () => __awaiter(this, void 0, void 0, function* () {
            const Reply = (0, responseBuilder_helper_1.default)(outers_1.StatusCodes.OK, "Server is healthy", {
                status: "ok",
                timestamp: new Date().toISOString(),
            });
            return Reply;
        }));
        // Available routes List
        fastify.get("/routes", (request, reply) => __awaiter(this, void 0, void 0, function* () {
            const Reply = (0, responseBuilder_helper_1.default)(outers_1.StatusCodes.OK, "Available routes", keys_1.AvailableRoutes);
            return reply.status(200).send(Reply);
        }));
        // Get Dashboard Stats
        fastify.get("/dashboard-stats", (request, reply) => __awaiter(this, void 0, void 0, function* () {
            return new Stats_controller_1.default(AxioDBInstance).getDashBoardStat();
        }));
        // Register the DB router
        fastify.register(DB_routes_1.default, {
            prefix: "/db",
            AxioDBInstance: AxioDBInstance, // Pass the AxioDB instance to the DB router
        });
        // Register Collection Router
        fastify.register(Collection_routes_1.default, {
            prefix: "/collection",
            AxioDBInstance: AxioDBInstance, // Pass the AxioDB instance to the Collection router
        });
        // Register Operation Router
        fastify.register(Operation_routes_1.default, {
            prefix: "/operation",
            AxioDBInstance: AxioDBInstance, // Pass the AxioDB instance to the Operation router
        });
        // Handle 404 Not Found
        fastify.setNotFoundHandler((request, reply) => {
            return reply
                .status(404)
                .send((0, responseBuilder_helper_1.default)(outers_1.StatusCodes.NOT_FOUND, `Route ${request.method}:${request.url} not found`));
        });
        done();
    });
}
