// outers/lib/Config/outer.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FunctionBased = exports.ClassBased = exports.Middleware = exports.Serve = exports.Retry = exports.Wait = exports.StatusCodes = exports.Console = void 0;
// Import Console Color Related Modules
const Console_log_1 = require("../Logs/Console.log"); // Import Console Module
// Import API Related Modules
const Dispatcher_1 = require("../API/Dispatcher"); // Import API Module
const Fetch_1 = require("../API/functions/Fetch"); // Import Fetch Module
// Import Response Related Modules
const Code_1 = require("../StatusCode/Code"); // Import StatusCode Module
const JSON_Response_1 = require("../Response/JSON-Response"); // Import JSON Response Module
const File_Response_1 = require("../Response/File-Response"); // Import File Response Module
const Render_Response_1 = __importDefault(require("../Response/Render-Response")); // Import Render Response Module
const JSON_Response_class_1 = __importDefault(require("../Response/Class/JSON-Response.class")); // Import JSON Responser Class
const File_Response_class_1 = __importDefault(require("../Response/Class/File-Response.class")); // Import File Responser Class
// Import Waiter Functions
const Waiter_function_1 = require("../Functions/Waiter.function"); // Waiter Functions
const Retry_function_1 = require("../Functions/Retry.function"); // Import Retry Function
// Import Random Generator Modules
const Base_1 = require("../UniqueGen/Base"); // Import UniqueGen Module
// Import Random Generator Module
const BooleanFunction_1 = __importDefault(require("../UniqueGen/Function/BooleanFunction")); // Random Boolean Generator
const MixedFunction_1 = __importDefault(require("../UniqueGen/Function/MixedFunction")); // Random Mixed Generator
const WordFunction_1 = __importDefault(require("../UniqueGen/Function/WordFunction")); // Random Word Generator
const SymbolFunction_1 = __importDefault(require("../UniqueGen/Function/SymbolFunction")); // Random Symbol Generator
const NumFunction_1 = __importDefault(require("../UniqueGen/Function/NumFunction")); // Random Number Generator
// Import JWT Manager Modules
const JWT_method_1 = __importDefault(require("../JWT/JWT.method")); // Import JWT Manager Module
// Import Cluster Creation Modules
const CreateClusterByFunction_method_1 = __importDefault(require("../Cluster/function/CreateClusterByFunction.method")); // Import Cluster Creation Module
const CreateClusterByClass_method_1 = __importDefault(require("../Cluster/class/CreateClusterByClass.method")); // Import Cluster Creation Module
// Import Encryption and Decryption Modules
const Dispatcher_2 = __importDefault(require("../Encryption - Decryption/Dispatcher")); // Import Crypto Module
// Import Storage Management Modules
const ShortStorage_storage_1 = __importDefault(require("../Storage Management/ShortStorage.storage")); // Import Short Storage Module
// Import Linux Command Execution Modules
const exec_command_1 = __importDefault(require("../command/exec.command")); // Linux Command Executor
const spawn_command_1 = __importDefault(require("../command/spawn.command")); // Linux Command Executor
// Import MongoDB Related Modules
const Connection_1 = __importDefault(require("../MongoSuper/Connection/Connection")); //  Mongo Super Module
// Import Git Cloning Modules
const Cloner_1 = __importDefault(require("../Git Cloner/Cloner")); // Import Git Cloner Module
// Import All Middlewares
const Base_middleware_1 = __importDefault(require("../Middlewares/Request IP Injector/Base.middleware")); // Import Inject IP Module
const Base_middleware_2 = __importDefault(require("../Middlewares/URL Controller/Base.middleware")); // Import URL Controller Module
const Base_middleware_3 = __importDefault(require("../Middlewares/IP Controller/Base.middleware")); // Import IP Controller Module
const Base_middleware_4 = __importDefault(require("../Middlewares/User Agent Controller/Base.middleware")); // Import User Access Controller Module
const Base_middleware_5 = __importDefault(require("../Middlewares/Request Controller/Base.middleware")); // Import Request Controller Module
const Base_middleware_6 = __importDefault(require("../Middlewares/JWT Validator/Base.middleware")); // Import JWT Validator Middleware
const Base_middleware_7 = __importDefault(require("../Middlewares/Request Logger/Base.middleware")); // Import Request Counter Middleware
// Import Functions Related Modules
const IP_Type_Checker_function_1 = __importDefault(require("../Functions/IP Type Checker.function")); // Import IP Type Checker Module
const Get_IP_Details_function_1 = __importDefault(require("../Functions/Get IP Details.function")); // Import Get IP Details Module
const Read_Request_Count_function_1 = __importDefault(require("../Functions/Read Request Count.function")); // Import Read Request Count Module
const createSocketClusterByClass_method_1 = __importDefault(require("../Cluster/class/createSocketClusterByClass.method"));
const CreateSocketClusterByFunction_method_1 = __importDefault(require("../Cluster/function/CreateSocketClusterByFunction.method"));
// Export Console Color Related Modules
exports.Console = Object.freeze({
    blue: Console_log_1.blue,
    bright: Console_log_1.bright,
    cyan: Console_log_1.cyan,
    dimmed: Console_log_1.dimmed,
    gray: Console_log_1.gray,
    green: Console_log_1.green,
    magenta: Console_log_1.magenta,
    red: Console_log_1.red,
    reverse: Console_log_1.reverse,
    underscore: Console_log_1.underscore,
    yellow: Console_log_1.yellow,
}); // Export Console Module
// Export HTTP Status Code Related Modules
exports.StatusCodes = Code_1.StatusCode; // Export StatusCode Module
// Export Waiter Modules
exports.Wait = Object.freeze({
    MS: Waiter_function_1.Ms,
    Seconds: Waiter_function_1.Seconds,
    Minutes: Waiter_function_1.Minutes,
    Hours: Waiter_function_1.Hours,
    Days: Waiter_function_1.Days,
    Weeks: Waiter_function_1.Weeks,
    Months: Waiter_function_1.Months,
    Centuries: Waiter_function_1.Centuries,
    Decades: Waiter_function_1.Decades,
    Millenia: Waiter_function_1.Millenia,
    Year: Waiter_function_1.Years,
});
// Export Retry Modules
exports.Retry = Object.freeze({
    MS: Retry_function_1.Ms,
    Seconds: Retry_function_1.Seconds,
    Minutes: Retry_function_1.Minutes,
    Hours: Retry_function_1.Hours,
    Days: Retry_function_1.Days,
    Weeks: Retry_function_1.Weeks,
    Months: Retry_function_1.Months,
    Centuries: Retry_function_1.Centuries,
    Decades: Retry_function_1.Decades,
    Millenia: Retry_function_1.Millenia,
    Year: Retry_function_1.Years,
});
// Export Direct Function Based HTTP Response Sender Modules
exports.Serve = Object.freeze({
    JSON: JSON_Response_1.JSONSendResponse,
    File: File_Response_1.SendFileResponse,
    Render: Render_Response_1.default,
});
// Export All Middlewares
exports.Middleware = Object.freeze({
    RequestInjectIP: Base_middleware_1.default, // Export IP Injector Module as Middleware
    URL_Controller: Base_middleware_2.default, // Export URL Controller Module as Middleware
    IPAccessController: Base_middleware_3.default, // Export IP Controller Module as Middleware
    User_AgentController: Base_middleware_4.default, // Export User Access Controller Module as Middleware
    MethodsController: Base_middleware_5.default, // Export Request Controller Module as Middleware
    JWTValidator: // Export Request Controller Module as Middleware
    Base_middleware_6.default, // Export JWT Validator Middleware
    RequestLogger: // Export JWT Validator Middleware
    Base_middleware_7.default, // Export Request Counter Middleware
}); // Export IP Injector Module as Middleware
// Export All Class based with Freeze
exports.ClassBased = Object.freeze({
    JWT_Manager: JWT_method_1.default, // Export JWT Manager Module
    MongoSuper: Connection_1.default,
    ClusterCreator: CreateClusterByClass_method_1.default, // CreateClusterByClass is used to create a cluster using the class based method
    SocketClusterCreator: createSocketClusterByClass_method_1.default, // CreateSocketCluster is used to create
    UniqueGenerator: Base_1.UniqueGen, // Export UniqueGen Module
    CryptoGraphy: Dispatcher_2.default, // Export Crypto Module
    API: Dispatcher_1.APiCall,
    Response: {
        JSON: JSON_Response_class_1.default, // Export Class Based JSON Response Sender Module
        File: File_Response_class_1.default, // Export Class Based File Response Sender Module
    }, // Export HTTP Response Sender Module
    Storage: {
        CreateNewShortStorage: ShortStorage_storage_1.default, // Export Short Storage Module
    }, // Export Storage Management Module
    Command: {
        Execute: exec_command_1.default, // Export Command Execution Module
        Spawn: spawn_command_1.default, // Export Command Execution Module
    }, // Export Command Related Module
    GitCloner: Cloner_1.default, // Export Git Cloning Module
}); // Export All Methods
// Export Functions with Freeze
exports.FunctionBased = Object.freeze({
    RandomGenerator: {
        Boolean: BooleanFunction_1.default, // Export Random Boolean Generator Module
        Mixed: MixedFunction_1.default, // Export Random Mixed Generator Module
        Number: NumFunction_1.default, // Export Random Number Generator Module
        Symbol: SymbolFunction_1.default, // Export Random Symbol Generator Module
        Word: WordFunction_1.default, // Export Random Word Generator Module
    }, // Export UniqueGen Module
    IP: {
        TypeCheck: IP_Type_Checker_function_1.default, // Export IP Type Checker Module
        Info: Get_IP_Details_function_1.default, // Export Get IP Details Function
    },
    ClusterCreator: CreateClusterByFunction_method_1.default, // CreateClusterByFunction is used to create a cluster using the function based method
    SocketClusterCreator: CreateSocketClusterByFunction_method_1.default, // CreateSocketCluster is used to create a socket cluster using the function based method
    API: {
        GET: Fetch_1.GetFetch, // Export Get Fetch Module
        POST: Fetch_1.PostFetch, // Export Post Fetch Module
        PUT: Fetch_1.PutFetch, // Export Put Fetch Module
        DELETE: Fetch_1.DeleteFetch, // Export Delete
    },
    GetRequestLog: Read_Request_Count_function_1.default, // Export Read Request Count Module
}); // Export
