// axiodb/lib/tcp/types/command.types.js
"use strict";
/**
 * Command types for TCP protocol
 * Defines all operations that can be performed over TCP connection
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CommandDocumentation = exports.CommandType = void 0;
var CommandType;
(function (CommandType) {
    // Connection Commands
    CommandType["PING"] = "PING";
    CommandType["DISCONNECT"] = "DISCONNECT";
    // Database Operations
    CommandType["CREATE_DB"] = "CREATE_DB";
    CommandType["DELETE_DB"] = "DELETE_DB";
    CommandType["DB_EXISTS"] = "DB_EXISTS";
    CommandType["GET_INSTANCE_INFO"] = "GET_INSTANCE_INFO";
    // Collection Operations
    CommandType["CREATE_COLLECTION"] = "CREATE_COLLECTION";
    CommandType["DELETE_COLLECTION"] = "DELETE_COLLECTION";
    CommandType["COLLECTION_EXISTS"] = "COLLECTION_EXISTS";
    CommandType["GET_COLLECTION_INFO"] = "GET_COLLECTION_INFO";
    // CRUD Operations
    CommandType["INSERT_DOCUMENT"] = "INSERT_DOCUMENT";
    CommandType["INSERT_MANY_DOCUMENTS"] = "INSERT_MANY_DOCUMENTS";
    CommandType["QUERY_DOCUMENTS"] = "QUERY_DOCUMENTS";
    CommandType["QUERY_BY_ID"] = "QUERY_BY_ID";
    CommandType["UPDATE_DOCUMENT_BY_ID"] = "UPDATE_DOCUMENT_BY_ID";
    CommandType["UPDATE_DOCUMENTS_BY_QUERY"] = "UPDATE_DOCUMENTS_BY_QUERY";
    CommandType["DELETE_DOCUMENT_BY_ID"] = "DELETE_DOCUMENT_BY_ID";
    CommandType["DELETE_DOCUMENTS_BY_QUERY"] = "DELETE_DOCUMENTS_BY_QUERY";
    CommandType["AGGREGATE"] = "AGGREGATE";
    CommandType["TOTAL_DOCUMENTS"] = "TOTAL_DOCUMENTS";
    // Index Operations
    CommandType["CREATE_INDEX"] = "CREATE_INDEX";
    CommandType["DROP_INDEX"] = "DROP_INDEX";
    // Transaction Operations (future)
    CommandType["BEGIN_TRANSACTION"] = "BEGIN_TRANSACTION";
    CommandType["COMMIT_TRANSACTION"] = "COMMIT_TRANSACTION";
    CommandType["ROLLBACK_TRANSACTION"] = "ROLLBACK_TRANSACTION";
})(CommandType || (exports.CommandType = CommandType = {}));
/**
 * Command documentation map for better error messages
 */
exports.CommandDocumentation = {
    [CommandType.PING]: 'Heartbeat ping to verify connection',
    [CommandType.DISCONNECT]: 'Gracefully disconnect from server',
    [CommandType.CREATE_DB]: 'Create a new database',
    [CommandType.DELETE_DB]: 'Delete an existing database',
    [CommandType.DB_EXISTS]: 'Check if database exists',
    [CommandType.GET_INSTANCE_INFO]: 'Get AxioDB instance information',
    [CommandType.CREATE_COLLECTION]: 'Create a new collection in database',
    [CommandType.DELETE_COLLECTION]: 'Delete a collection from database',
    [CommandType.COLLECTION_EXISTS]: 'Check if collection exists',
    [CommandType.GET_COLLECTION_INFO]: 'Get collection metadata',
    [CommandType.INSERT_DOCUMENT]: 'Insert single document into collection',
    [CommandType.INSERT_MANY_DOCUMENTS]: 'Insert multiple documents into collection',
    [CommandType.QUERY_DOCUMENTS]: 'Query documents with filters and options',
    [CommandType.QUERY_BY_ID]: 'Query document by ID',
    [CommandType.UPDATE_DOCUMENT_BY_ID]: 'Update document by ID',
    [CommandType.UPDATE_DOCUMENTS_BY_QUERY]: 'Update documents matching query',
    [CommandType.DELETE_DOCUMENT_BY_ID]: 'Delete document by ID',
    [CommandType.DELETE_DOCUMENTS_BY_QUERY]: 'Delete documents matching query',
    [CommandType.AGGREGATE]: 'Run aggregation pipeline',
    [CommandType.TOTAL_DOCUMENTS]: 'Get total document count',
    [CommandType.CREATE_INDEX]: 'Create index on collection fields',
    [CommandType.DROP_INDEX]: 'Drop existing index',
    [CommandType.BEGIN_TRANSACTION]: 'Begin database transaction',
    [CommandType.COMMIT_TRANSACTION]: 'Commit database transaction',
    [CommandType.ROLLBACK_TRANSACTION]: 'Rollback database transaction',
};
